### ISSUE TITLE
Order details and previews showing tags

### ISSUE THREAD

**Issue by MarioP175** [qtranslate/qtranslate-xt#820]
In Woocommerce->Orders when you click on eye icon to preview order or just on order number field item is not translated.

Tried to ad __() to class-wc-admin-list-table-orders.php. Did not help. Any ideas?

**Comment by MarioP175** [comment]
Actually everything in dashboard also is with tags and not translated

**Comment by herrvigg** [comment]
Can you specify exactly which tags? Maybe give a screenshot.

One field not translated is the "notes". It should be easy to fix for the order page but on the preview it might be more difficult. Anything that has to do with dynamic javascript is more problematic, unless it triggers an API REST call that can be handled on the server.

For the dashboard can you also provide a screenshot?

**Comment by herrvigg** [comment]
> Tried to ad __() to class-wc-admin-list-table-orders.php. Did not help.

Even if it worked, adding `__()` is not a viable solution because it goes against the logic of this function, which is only meant for basic translations of static labels, not custom/dynamic translations as qTranslate. I'm quite sure the WooCommerce developers will refuse to encapsulate with such functions. The only way to handle this properly is through hooks and actions. If there is no such callbacks you might ask them to add those instead but it has to make some sense in WC from a functional point of view.

**Comment by MarioP175** [comment]
[image: image.png]
[image: изображение.png]



[image: изображение.png]


ср, 22 апр. 2020 г. в 11:03, Herr Vigg <notifications@github.com>:

**Comment by herrvigg** [comment]
The attachments are missing. I can't see them on github and not even in the email notification either.

Can you upload them directly on github? https://github.com/qtranslate/qtranslate-xt/issues/820

It's very convenient, you can copy-paste an image directly from the clipboard. Possibly crop as much you as can to reduce the size.

**Comment by MarioP175** [comment]
<img width="1017" alt="изображение" src="https://user-images.githubusercontent.com/53692819/79972655-35bca780-848e-11ea-9fb9-0674e3501f84.png">
<img width="610" alt="изображение" src="https://user-images.githubusercontent.com/53692819/79972729-4bca6800-848e-11ea-908f-57e0fa05e63d.png">

**Comment by herrvigg** [comment]
Thanks it helps, so these should be the product titles. Just wondering, is this a new problem with WC version 4 or did it happen already before?

**Comment by MarioP175** [comment]
Don't know. Just installed

ср, 22 апр. 2020 г. в 11:55, Herr Vigg <notifications@github.com>:

**Comment by herrvigg** [comment]
The translation of the items in the order details is now fixed with #827 

The problem with the order preview and the dashboard remain.

**Comment by herrvigg** [comment]
Those two are more difficult to handle, they can't be handled by standard JSON config because it's AJAX and REST requests:
- order preview: `/wp-admin/admin-ajax.php?order_id=24319&action=woocommerce_get_order_details`
- dashboard: `/wp-json/wc-analytics/leaderboards`

**Comment by pjrobertson** [comment]
I can confirm that the orders list page still has several places not translated, and it's more tricky because they're AJAX requests. It seems like @herrvigg you've already debugged, but here's a screenshots anyway, after clicking the 'eye' icon. WooCommerce 3.9.3, WP5.2

![Screenshot 2020-04-30 at 06 57 07](https://user-images.githubusercontent.com/150431/80654909-fc002800-8aaf-11ea-9c5e-a1701f048e5f.jpg)

**Comment by herrvigg** [comment]
In theory it should be possible to fix. But in practice it can be very problematic. The big challenge with REST and AJAX requests is that you don't know all the context in which the data is used. If it's only for display sure you should translate. But if the data is used in a state and sent back later to be written in the DB it will erase all the translations which is not something you want obviously. And you don't necessarily know this from the API itself. This is the main idea behind REST from a conceptual point of view. Ideally the context should be part of the call itself but it goes often beyond the regular use of such APIs which don't really care about the translations.

**Comment by pjrobertson** [comment]
I agree this is tricky for AJAX calls. One solution could be to do all the
changes client side. Add a listener to AJAX calls or could even add a
'changed' listener to specific elements. If their value changes, then
re-run the client-side code to search for elements to translate.
